src/wp-includes/class-wp-block-scanner.php

```diff
@@ -0,0 +1,1244 @@
+	 * or the length of the source text if finished parsing.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $next_token_at = 0;
+
+	/**
+	 * Byte offset where namespace span begins.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $namespace_at = 0;
+
+	/**
+	 * Byte length of namespace span, or `0` if implicitly in the "core" namespace.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $namespace_length = 0;
+
+	/**
+	 * Byte offset where block name span begins.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $name_at = 0;
+
+	/**
+	 * Byte length of block name span.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $name_length = 0;
+
+	/**
+	 * Whether the delimiter contains the block-closing flag.
+	 *
+	 * This may be erroneous if present within a void block,
+	 * therefore the {@see self::has_closing_flag} can be used by
+	 * calling code to perform appropriate error-handling.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var bool
+	 */
+	private $has_closing_flag = false;
+
+	/**
+	 * Byte offset where JSON attributes span begins.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $json_at;
+
+	/**
+	 * Byte length of JSON attributes span, or `0` if none are present.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var int
+	 */
+	private $json_length = 0;
+
+	/**
+	 * Internal parser state, differentiating whether the instance is currently matched,
+	 * on an implicit freeform node, in error, or ready to begin parsing.
+	 *
+	 * @see self::READY
+	 * @see self::MATCHED
+	 * @see self::HTML_OPEN
+	 * @see self::HTML_CLOSE
+	 * @see self::INCOMPLETE_INPUT
+	 * @see self::COMPLETE
+	 *
+	 * @since 6.9.0}
+	 *
+	 * @var string
+	 */
+	protected $state = self::READY;
+
+	/**
+	 * Indicates what kind of block comment delimiter this represents.
+	 *
+	 * One of:
+	 *
+	 *  - {@see static::OPENER} If the delimiter is opening a block.
+	 *  - {@see static::CLOSER} If the delimiter is closing an open block.
+	 *  - {@see static::VOID}   If the delimiter represents a void block with no inner content.
+	 *
+	 * If a parsed comment delimiter contains both the closing and the void
+	 * flags then it will be interpreted as a void block to match the behavior
+	 * of the official block parser, however, this is a mistake and probably
+	 * the block ought to close an open block of the same name, if one is open.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @var string
+	 */
+	private $type;
+
+	/**
+	 * Creates a new block scanner.
+	 *
+	 * Example:
+	 *
+	 *     $scanner = WP_Block_Scanner::create( $html );
+	 *     while ( $scanner->next_delimiter() ) {
+	 *          if ( $scanner->opens_block( 'core/image' ) ) {
+	 *             echo "Found an image!\n";
+	 *         }
+	 *     }
+	 *
+	 * @see self::next_delimiter
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param string $source_text Input document potentially containing block content.
+	 * @return ?self Created block scanner, if successfully created.
```

Comment:
In the HTML processor we went with `static|null`. This indicates that an instance of the subclass is returned in case of extension:

```suggestion
	 * @return static|null Created block scanner, if successfully created.
```

https://github.com/WordPress/wordpress-develop/blob/2da7a592d643939e0925b821ae4cc2eed29b0db5/src/wp-includes/html-api/class-wp-html-processor.php#L345-L360

---

The `?static` return type annotation does not appear to be supported until PHP 8.0.

Reply:
hm. I seemed to have confused `self` and `static` once again, as I thought that `static` locked us in to the lexical scope, but `self` was more dynamic for subclassing. confusing. 🤷‍♂️

Reply:
Does a `@return` even make sense here, since constructors don't return values?

Reply:
this method is now gone.